contracts/PseudoRandomGenerator.cdc

```diff
@@ -3,19 +3,65 @@ access(all) contract PseudoRandomGenerator {
     access(all) let StoragePath: StoragePath
     access(all) let PublicPath: PublicPath
 
+    access(all) event NextUInt64(id: UInt64, value: UInt64)
+
+    /// While not limited to 128 bits of state, this PRG is largely informed by XORShift128+
+    ///
     access(all) resource PRG {
         access(all) let sourceOfRandomness: [UInt8]
-        access(all) let salt: UInt64
+        access(all) let salt: Word64
+        
+        /// The states below are of type Word64 to prevent overflow/underflow
+        //
+        access(all) var state0: Word64
+        access(all) var state1: Word64
         
         init(sourceOfRandomness: [UInt8], salt: UInt64) {
+            pre {
+                sourceOfRandomness.length == 32: "Expecting 32 bytes"
+            }
             self.sourceOfRandomness = sourceOfRandomness
-            self.salt = salt
+            self.salt = Word64(salt)
+
+            // Convert the seed bytes to two Word64 values for state initialization
+            let segment0 = PseudoRandomGenerator.bytesToWord64(bytes: sourceOfRandomness, start: 0)
+            let segment1 = PseudoRandomGenerator.bytesToWord64(bytes: sourceOfRandomness, start: 8)
+            let segment2 = PseudoRandomGenerator.bytesToWord64(bytes: sourceOfRandomness, start: 16)
+            let segment3 = PseudoRandomGenerator.bytesToWord64(bytes: sourceOfRandomness, start: 24)
+
+            self.state0 = segment0 ^ segment1
+            self.state1 = segment2 ^ segment3
         }
 
-        // TODO: replace unsafeRandom with PRG implementation
         access(all) fun nextUInt64(): UInt64 {
-            return unsafeRandom()
+            var a: Word64 = self.state0
+            let b: Word64 = self.state1
+
+            self.state0 = b
+            a = a ^ (a << 23) // a
+            a = a ^ (a >> 17) // b
+            a = b ^ (b >> 5) // c
+            self.state1 = a
+
+            let randUInt64: UInt64 = UInt64((a + b) ^ self.salt)
```

Comment:
as per https://github.com/onflow/random-coin-toss/issues/5#issuecomment-1793277245, 
```suggestion
            let randUInt64: UInt64 = UInt64(a + b)
```
Btw I believe `a+b` is done first as `Word64` and then the cast into `UInt64` happens right?

Reply:
Yes, that's right